golem-component-service-base/src/repo/component.rs

```diff
@@ -0,0 +1,467 @@
+        };
+        Ok(Component {
+            component_name: ComponentName(value.name),
+            component_size: value.size as u64,
+            metadata,
+            versioned_component_id,
+            user_component_id,
+            protected_component_id,
+        })
+    }
+}
+
+impl From<ComponentRecord> for VersionedComponentId {
+    fn from(value: ComponentRecord) -> Self {
+        VersionedComponentId {
+            component_id: ComponentId(value.component_id),
+            version: value.version as u64,
+        }
+    }
+}
+
+impl ComponentRecord {
+    pub fn new<Namespace: Display>(
+        namespace: Namespace,
+        component: Component,
+    ) -> Result<Self, String> {
+        let metadata = record_metadata_serde::serialize(&component.metadata)?;
+        Ok(Self {
+            namespace: namespace.to_string(),
+            component_id: component.versioned_component_id.component_id.0,
+            name: component.component_name.0,
+            size: component.component_size as i32,
+            version: component.versioned_component_id.version as i64,
+            user_component: component.versioned_component_id.slug(),
+            protected_component: component.protected_component_id.slug(),
+            protector_version: None,
+            metadata: metadata.into(),
+        })
+    }
+}
+
+#[async_trait]
+pub trait ComponentRepo {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError>;
+
+    async fn get(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_all(&self, namespace: &str) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_latest_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_version(
+        &self,
+        namespace: &str,
+        component_id: &Uuid,
+        version: u64,
+    ) -> Result<Option<ComponentRecord>, RepoError>;
+
+    async fn get_by_name(
+        &self,
+        namespace: &str,
+        name: &str,
+    ) -> Result<Vec<ComponentRecord>, RepoError>;
+
+    async fn get_ids_by_name(&self, namespace: &str, name: &str) -> Result<Vec<Uuid>, RepoError>;
+
+    async fn get_namespaces(&self, component_id: &Uuid) -> Result<Vec<(String, i64)>, RepoError>;
+
+    async fn delete(&self, namespace: &str, component_id: &Uuid) -> Result<(), RepoError>;
+}
+
+pub struct DbComponentRepo<DB: Database> {
+    db_pool: Arc<Pool<DB>>,
+}
+
+impl<DB: Database> DbComponentRepo<DB> {
+    pub fn new(db_pool: Arc<Pool<DB>>) -> Self {
+        Self { db_pool }
+    }
+}
+
+#[async_trait]
+impl ComponentRepo for DbComponentRepo<sqlx::Sqlite> {
+    async fn upsert(&self, component: &ComponentRecord) -> Result<(), RepoError> {
+        let mut transaction = self.db_pool.begin().await?;
+
+        let result = sqlx::query(
```

Comment:
This is not the method result. Please consider renaming it to something like `same_id` or inlining it.